runtime/near-vm-runner/src/wasmer2_runner.rs

```diff
@@ -266,26 +272,106 @@ impl Wasmer2VM {
+                    None => None,
+                    Some(CompiledContract::CompileModuleError(err)) => return Ok(Err(err)),
+                    Some(CompiledContract::Code(serialized_module)) => {
+                        unsafe {
+                            // (UN-)SAFETY: the `serialized_module` must have been produced by a prior call to
+                            // `serialize`.
+                            //
+                            // In practice this is not necessarily true. One could have forgotten to change the
+                            // cache key when upgrading the version of the wasmer library or the database could
+                            // have had its data corrupted while at rest.
+                            //
+                            // There should definitely be some validation in wasmer to ensure we load what we think
+                            // we load.
+                            let executable =
+                                UniversalExecutableRef::deserialize(&serialized_module)
+                                    .map_err(|_| CacheError::DeserializationError)?;
+                            let artifact = self
+                                .engine
+                                .load_universal_executable_ref(&executable)
+                                .map(Arc::new)
+                                .map_err(|_| CacheError::DeserializationError)?;
+                            Some(artifact)
+                        }
+                    }
+                };
+
+                let artifact: VMArtifact = match stored_artifact {
+                    Some(it) => it,
+                    None => {
+                        let executable_or_error = self.compile_uncached(code);
+                        if let Some(cache) = cache {
+                            let record = match &executable_or_error {
+                                Ok(executable) => {
+                                    let code = executable.serialize().map_err(|_e| {
+                                        CacheError::SerializationError { hash: key.0 }
+                                    })?;
+                                    CompiledContract::Code(code)
+                                }
+                                Err(err) => CompiledContract::CompileModuleError(err.clone()),
+                            };
+                            cache
+                                .put(&key, record)
+                                .map_err(|_: io::Error| CacheError::WriteError)?;
+                        }
+                        match executable_or_error {
+                            Ok(executable) => self
+                                .engine
+                                .load_universal_executable(&executable)
+                                .map(Arc::new)
+                                .map_err(|_| CacheError::DeserializationError)?,
```

Comment:
This error variant seems like it would be quite confusing if one saw it as a result of OOM or some other failure in resource management.

Reply:
Yeah, will look into this. This is somewhat pre-existing

https://github.com/near/nearcore/pull/7851/files#diff-9e6bfc7fd517c992e3a50e28e15995fca1a8c5b6f2b8b3cafc775c91317c59bbL307

and i didn't want to change the shape of errors here, but perhaps we should. 

"LoadingError" is a genuine error condition here, so this needs to return something other than CacheError. Before, we just panicked or swallowed these errors.